[MHDP-7] Added new roles

prnvshrn opened the pull request:
- Added option to add a new position record which has the <b>new roles (Corporation Member,Trustee,Friday Evening Lecturer) </b> and also, the optional <b>start and end dates</b>
- The position information can be updated and deleted
- The information of the new position records is also reflected in the <b>people</b> API response

jdamerow commented on browser/views.py:
```diff
@@ -687,3 +687,109 @@ def delete_investigator_record(request, person_id, research_id):
+                end_month = end_date.split('/')[0]
+                end_year = end_date.split('/')[2]
+                end_day = end_date.split('/')[1]
+                end_date = end_year + '-' + end_month + '-' + end_day
```
if you use a datefield in the form for those fields, then it should already be validated and converted, correctly, shouldn't it? This seems very hacky.

prnvshrn replied:
Removed this date conversion python code and instead, changed the date format of the html date picker. Date picker returns the correct format which need not be converted again in the python side.

jdamerow replied:
how about validation? shouldn't the datefield do that for you? what kind of field is it right now? String or Date?

prnvshrn replied:
The field being returned to back-end is of string data type

jdamerow commented on browser/views.py:
```diff
@@ -687,3 +687,109 @@ def delete_investigator_record(request, person_id, research_id):
+        'form': form,
+        'person': person,
+        'type': 'edit_position'
+    }
```
duplicate code. I'm pretty sure you can do this in the add method.

prnvshrn replied:
I have combined add and edit methods into a single entity.

prnvshrn commented on browser/models.py:
```diff
@@ -506,3 +510,37 @@ def _history_user(self):
+                rep += u', '
+            rep += u'Subject: %s' % self.subject
+        if self.subject or self.role:
+            rep += u', in %s' % self.year
```
I have removed the commas

prnvshrn commented on browser/models.py:
```diff
@@ -506,3 +510,37 @@ def _history_user(self):
+    person = models.ForeignKey('Person')
+    subject = models.CharField(max_length=255, blank=True)
+    role_choices = (('Corporation Member','Corporation Member'), ('Trustee','Trustee'),('Friday Evening Lecturer', 'Friday Evening Lecturer'))
+    role = models.CharField(max_length=255, blank=False,choices=role_choices)
```
Changed the actual values of the tuples to constants

prnvshrn commented on browser/templates/browser/person.html:
```diff
@@ -261,11 +323,261 @@ <h3>Locations</h3>
+                            <a class=" btn btn-sm fa fa-edit"
+                               aria-hidden="true"
+                               id="edit_position"
+                               href="{% url "add-position" person.id position.id %}">
```
I edited the URL name from 'add-position' to only 'position'

prnvshrn commented on browser/views.py:
```diff
@@ -687,3 +686,64 @@ def delete_investigator_record(request, person_id, research_id):
+                start_date = request.POST.get("start_date")
+
+            if '-' in request.POST.get("end_date"):
+                end_date = request.POST.get("end_date")
```
I am using a datefield in model but was not utilizing it properly. So, I have removed the above code for retrieving start and end dated and instead now using the following 
```
start_date=form.cleaned_data.get('start_date'),
end_date=form.cleaned_data.get('end_date')
```

prnvshrn commented on browser/views.py:
```diff
@@ -687,3 +686,64 @@ def delete_investigator_record(request, person_id, research_id):
+                position.year = form.cleaned_data.get('year')
+                position.start_date = start_date
+                position.end_date = end_date
+                position.save()
```
Edited the code to update and create database through dictionaries, like the following
```
Position.objects.create(
                    person_id=person_id,
                    changed_by=request.user,
                    **form.cleaned_data
                )
```

prnvshrn commented on browser/views.py:
```diff
@@ -687,3 +686,64 @@ def delete_investigator_record(request, person_id, research_id):
+    context = {
+        'form': form,
+        'person': person,
+        'type': 'create_position'
```
Removed type variable. Position id is used to determine if operation is update or create.

prnvshrn commented on browser/views.py:
```diff
@@ -687,3 +686,64 @@ def delete_investigator_record(request, person_id, research_id):
+    }
+
+    # If position_id = 0 then, we are dealing with adding a new position request
+    if position_id != "0":
```
position_id is compared with None instead of '0'

jdamerow commented on browser/models.py:
```diff
@@ -506,3 +510,41 @@ def _history_user(self):
+    CORP = 'Corporation Member'
+    TRUSTEE = 'Trustee'
+    FEL = 'Friday Evening Lecturer'
+    role_choices = ((CORP, 'Corporation Member'), (TRUSTEE, 'Trustee'),
```
Please see the django documentation for an example of this. 'Corporation Member' is not a good 'key'. way to easy to be confused with the label.

jdamerow commented on browser/templates/browser/position.html:
```diff
@@ -0,0 +1,164 @@
+
+<div class="container">
+    <p class="text-info">
+        {% if position_id == undefined or position_id == null %}
```
can't you just say ```if position_id```?

prnvshrn replied:
Changed the condition to only ```if position_id```

jdamerow commented on browser/templatetags/app_filters.py:
```diff
@@ -141,3 +141,11 @@ def get_affiliates(institution):
+@register.filter
+def get_positions(person, role_type):
+    if role_type == 'All':
+        return Position.objects.filter(person=person).distinct('year', 'id').order_by('year', 'id')
```
are there objects that are returned twice or why do you do an distinct on year and id?

prnvshrn replied:
The <b>distinct</b> part was not required so I have removed it from the query. Also I have removed <b>id</b> from the <b>order-by</b> part of the query.

jdamerow commented on browser/views.py:
```diff
@@ -687,3 +686,47 @@ def delete_investigator_record(request, person_id, research_id):
+        'person': person,
+    }
+
+    if position_id is not None:
```
```if position_id: ```~~is not None~~

prnvshrn replied:
Removed ``` is not None ```

hashknot replied:
Can `position_id` be 0? If yes, `if position_id:` might not behave as expected.

jdamerow commented on browser/views.py:
```diff
@@ -687,3 +686,47 @@ def delete_investigator_record(request, person_id, research_id):
+        })
+
+    if request.method == 'POST':
+        form = PositionForm(request.POST, instance=person)
```
if there is a position id, then ```form``` is already set, isn't it?

prnvshrn replied:
Yes, the form is already set if there is a position_id .  So, I have removed ```instance=person```

jdamerow commented on mbl/urls.py:
```diff
@@ -36,6 +36,9 @@
     url(r'^person/(?P<person_id>[0-9]+)/investigator/(?P<research_id>[0-9]+)/edit/$',views.edit_investigator_record,name='edit-investigator'),
     url(r'^person/(?P<person_id>[0-9]+)/investigator/(?P<research_id>[0-9]+)/delete/$', views.delete_investigator_record, name='delete-investigator'),
+    url(r'^person/(?P<person_id>[0-9]+)/position/add/$', views.position, {'position_id': None}, name='add-position'),
+    url(r'^person/(?P<person_id>[0-9]+)/position/(?P<position_id>[0-9]+)/edit/$', views.position, name='edit-position'),
```
since add and edit both point to the same view function, why two urls?

prnvshrn replied:
I was not able to pass a none object from the html template to the django view. So, I am passing the none object from the urls file itself. That is why I have used an additional url here.

hashknot replied:
Why not different view functions? You won't need to add/edit in the same function based on `position_id` being `None`.

prnvshrn replied:
@hashknot Earlier there were two views to handle these cases however, there was a lot of redundant code between them. That's why only one function is being used to add and edit.

prnvshrn commented:
I have made the following revisions -
1. Changed edit and add urls into one url 
2. Changed the keys in role_choices tuples in models file
3. Added datable for displaying data in Position tab and also, added a dropdown for sorting based on type of role

hashknot commented on browser/models.py:
```diff
@@ -506,3 +510,41 @@ def _history_user(self):
+    FRIDAY_EVENING_LECTURER = 'FEL'
+    role_choices = ((CORPORATION_MEMBER, 'Corporation Member'), (TRUSTEE, 'Trustee'),
+                    (FRIDAY_EVENING_LECTURER, 'Friday Evening Lecturer'))
+    role = models.CharField(max_length=255, blank=False, choices=role_choices, default=CORPORATION_MEMBER)
```
A word of caution: `blank=False` only applies to forms. If you want the field nullable in the database, make sure you add `null=True` here. Same goes for the rest of the fields.

If you do add `null=True`, squash all the new migrations into one to allow this change specific migrations to be undone.

hashknot replied:
Also, is `max_length=255` actually necessary for `role` when there are only three possible values with maximum string length being 3?

prnvshrn replied:
I did not want the field to be nullable in the back-end. So, I have used ```blank=False``` to validate in the form itself.
The max_length is 255 so that in the future we can add roles of varied lengths without making changes in the models.

jdamerow replied:
Since there are keys it is highly unlikely that we will have keys that are 255 in the future. 3 is probably enough even if we add a lot more options (which is unlikely).

jdamerow replied:
this is still set to 255

prnvshrn replied:
Changed the length of role to 3.

jdamerow commented on browser/migrations/0019_auto_20171103_2147.py:
```diff
@@ -0,0 +1,49 @@
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
```
seems like this and the previous migration create position and delete it again. Just remove the migrations if that's the case.

prnvshrn replied:
I removed all of the previous migrations

jdamerow commented on browser/views.py:
```diff
@@ -687,3 +686,47 @@ def delete_investigator_record(request, person_id, research_id):
+    context = {
+        'form': form,
+        'person': person,
+        'position_id': None
```
shouldn't position_id be position_id (what is passed into the method)?

prnvshrn replied:
Yes, the position_id should be set to position_id. I have changed the above code to ```'position_id' : position_id```

jdamerow commented on browser/views.py:
```diff
@@ -687,3 +686,47 @@ def delete_investigator_record(request, person_id, research_id):
+            else:
+                Position.objects.filter(id=position_id).update(**form.cleaned_data)
+            return HttpResponseRedirect(reverse('person', args=(person.id,)))
+    return render(request, template, context)
```
if the form is not valid, we still need it in the context, don't we? to be able to show errors.

prnvshrn replied:
The form is in the context when we are initially calling the position function. So, the errors in the form are displayed to the user.

jdamerow replied:
but if there is a position id (so position is edited), then the form would be overwritten by line 704 and 709, doesn't it?

prnvshrn replied:
When we are making a POST request , a call is made to fetch the form at line 713 - <br>``` form = PositionForm(request.POST) ```. <br>This gets the value which the user wants to update.
The lines 704 and 709 do set the form to the initial value. But , the above code retrieves the correct values from the POST request that needs to be updated.

hashknot replied:
You are not returning a validated form (with errors) in the context. As a result, the template won't be able to display _why_ the POST request didn't go through.

jdamerow replied:
besides what Jitesh said, this is also unnecessarily very confusing. Just have a GET and a POST block, initialize the forms in them as needed (either with object, without, or from the POST request) and then update the context accordingly (in each block). There is no reason to initialize the form in line 695 if there is just one case where an empty form is needed.

prnvshrn replied:
I have changed the implementation of the form. There is a GET and POST block, and the forms are being initialized within these blocks.

jdamerow commented:
why did you remove **all** migrations? that seems wrong and seems like it'll cause havoc when deployed.

jdamerow commented on browser/views.py:
```diff
@@ -687,3 +686,50 @@ def delete_investigator_record(request, person_id, research_id):
+            form = PositionForm()
+            context.update({
+                'form': form
+            })
```
update is duplicate code; move outside of if/else statement.

prnvshrn replied:
Moved the duplicate code out of the if / else statement

jdamerow commented:
It should be possible to type in a date in these fields as well (if I want to enter e.g. 01-01-1980, I don't want to have to use the dropdowns):
![image](https://user-images.githubusercontent.com/8881141/37684268-bbe196b4-2c4c-11e8-97f9-b19b8fef0ac0.png)

jdamerow commented:
This table should look like all the others:
![image](https://user-images.githubusercontent.com/8881141/37684356-02a6daf0-2c4d-11e8-9a2f-cd6bf64231ae.png)

jdamerow commented:
after adding a new position, the user should be redirected to the positions table.

jdamerow commented:
The icon for adding a new position is a little confusing.
![image](https://user-images.githubusercontent.com/8881141/37684800-6920b976-2c4e-11e8-8046-2d19e31b6bd8.png)
Since we have also people, this is easily confused with a person record. Let's maybe use the id-card or id-badge.

prnvshrn commented:
I have made the following changes -
1. Put redundant code outside of if-else statement
2. Made date type-able (Working on validating it in client side)
3. Removed the datable and used the bootstrap table to make it similar to other tables
4. Changed the position icon to id-badge

jdamerow commented:
@prnvshrn did you mean to reopen this PR?

prnvshrn commented:
When a new position is created or an existing is edited, the position tab will open in <b>Persons</b> page

jdamerow commented:
please resolve conflicts

prnvshrn commented:
Removed the merge conflicts

jdamerow commented:
as discussed, the global variable needs to go

prnvshrn commented:
Removed the global variable. The information of tab is passed through url parameter.

jdamerow commented on browser/templates/browser/person.html:
```diff
@@ -261,11 +308,133 @@ <h3>Locations</h3>
+            </div>
+            <div class="container">
+            <label for="dropdownTest" >Sort by Role:</label>
+            <select id="dropdownTest" name="cars">
```
why is this called 'dropdownTest'? Why is the name 'cars'?

prnvshrn replied:
Changed the variable names

jdamerow commented on browser/templatetags/app_filters.py:
```diff
@@ -141,3 +141,16 @@ def get_affiliates(institution):
+    list_of_roles = []
+    for i in Position.role_choices:
+        list_of_roles.append(i[1])
+    return list_of_roles
```
using list comprehension this should be a one-liner

prnvshrn replied:
Reduced the above code to a single line

jdamerow commented on browser/views.py:
```diff
@@ -686,3 +685,51 @@ def delete_investigator_record(request, person_id, research_id):
+                )
+            else:
+                Position.objects.filter(id=position_id).update(**form.cleaned_data)
+            return HttpResponseRedirect(reverse('person', args=(person.id,),)+'?person-positions')
```
this is not proper url format. Parameters are not passed like that.

prnvshrn replied:
Changed passing of url parameters to the following:
```
('person', args=(person.id,),)+'?tab=person-positions')
```

jdamerow commented on browser/templates/browser/person.html:
```diff
@@ -2,12 +2,48 @@
+            if(typeof person_url_params[1] !== 'undefined' && typeof person_url_params[1] !== 'null' )
+            {
+                $('.nav-tabs a[href="#' + person_url_params[1] + '"]').tab('show');
+            }
```
if you use proper parameters in the url, you can simply get a parameter from the request using django's template syntax.

jdamerow commented on browser/templates/browser/person.html:
```diff
@@ -2,12 +2,47 @@
+            if(typeof person_url_params[1] !== 'undefined' && typeof person_url_params[1] !== 'null' )
+            {
+                $('.nav-tabs a[href="#' + person_url_params[1] + '"]').tab('show');
+            }
```
what about my previous comment here?

prnvshrn replied:
Sorry about that. I am now using Django's request parameter to get information of tab parameter as following.
```
var tab = "{{request.GET.tab}}";
            if(typeof tab !== 'undefined' && typeof tab !== 'null' )
            {
                $('.nav-tabs a[href="#' + tab + '"]').tab('show');
            }
```

jdamerow replied:
just use django template 'if' and get rid of the ugly javascript.

prnvshrn replied:
Changed to Django if condition as following:
```
{% if request.GET.tab %}
$('.nav-tabs a[href="#' + "{{request.GET.tab}}" + '"]').tab('show');
{% endif %}
```
